wiring/inc/spark_wiring_system.h

```diff
@@ -148,8 +148,18 @@ class SystemClass {
 
     inline int enableFeature(HAL_Feature feature)
     {
+        /* PRE-ENABLE TASKS */
+        bool init_sram = false;
+        // Initialize SRAM on first use of retained memory
+        if (feature==FEATURE_RETAINED_MEMORY && !HAL_Feature_Get(FEATURE_RETAINED_MEMORY)) {
+            init_sram = true;
+        }
+
+        /* ENABLE FEATURE */
         int result = HAL_Feature_Set(feature, true);
-        if (feature==FEATURE_RETAINED_MEMORY && !HAL_Feature_Get(FEATURE_WARM_START)) {
+
+        /* POST-ENABLE TASKS */
+        if (feature==FEATURE_RETAINED_MEMORY && init_sram) {
```

Comment:
toggle a gpio and measure the time.

On Fri, Oct 9, 2015 at 2:17 PM, Technobly <notifications@github.com> wrote:

> In wiring/inc/spark_wiring_system.h
> <https://github.com/spark/firmware/pull/664#discussion_r41667856>:
>
> >          int result = HAL_Feature_Set(feature, true);
> > -        if (feature==FEATURE_RETAINED_MEMORY && !HAL_Feature_Get(FEATURE_WARM_START)) {
> > +
> > +        /* POST-ENABLE TASKS */
> > +        if (feature==FEATURE_RETAINED_MEMORY && init_sram) {
>
> I could see the only reason to do that would be to ensure SRAM is
> initialized when FEATURE_RETAINED_MEMORY is not enabled (for when you might
> want an extra 4KB of RAM that is not retained after power loss or reset). I
> guess the question is, how long does 4KB of memcpy() take? Or does it even
> initialize more than is actually used? I'd prefer not to tack on more boot
> time by default, regardless of feature use. If we assumed 42ns per byte
> that's about 176us... not bad I guess but is that a realistic estimate?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark/firmware/pull/664/files#r41667856>.
>



-- 
Andy